engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrowArrayBuilder.java

```diff
@@ -0,0 +1,242 @@
+package org.enso.interpreter.runtime.data.vector;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.Warning;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+import org.enso.interpreter.runtime.error.WithWarnings;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(
+    pkg = "mutable",
+    stdlibName = "Standard.Base.Data.ArrowArrayBuilder.ArrowArrayBuilder",
+    containsValues = false)
+public abstract class ArrowArrayBuilder implements EnsoObject {
+
+  protected boolean sealed = false;
+
+  public void seal() {
+    sealed = true;
+  }
+
+  @ExportMessage
+  boolean hasArrayElements() {
+    return true;
+  }
+
+  @ExportMessage
+  boolean isArrayElementModifiable(long index) {
+    return true;
+  }
+
+  @ExportMessage
+  boolean isArrayElementReadable(long index) {
+    return true;
+  }
+
+  @ExportMessage
+  long getArraySize() {
+    throw CompilerDirectives.shouldNotReachHere();
+  }
+
+  @ExportMessage
+  Object readArrayElement(long index)
+      throws UnsupportedMessageException, InvalidArrayIndexException {
+    throw CompilerDirectives.shouldNotReachHere();
+  }
+
+  @ExportMessage
+  final void writeArrayElement(long index, Object value) throws UnsupportedMessageException {
+    throw UnsupportedMessageException.create();
+  }
+
+  @ExportMessage
+  boolean isArrayElementInsertable(long index) {
+    throw CompilerDirectives.shouldNotReachHere();
```

Comment:
Why we `shouldNotReachHere`?

Reply:
It's overwritten by the concrete implementation